File: src/main/scala-2.11/app/gui/controller/impl/AppController.scala
package app.gui.controller.impl


import java.io.{File, FileInputStream}

import app.gui.controller.{Controller, TAppController}
import app.gui.model.{AppModel, Model}
import app.util.{Constants, Util}

import scalafx.scene.canvas.GraphicsContext
import scalafx.scene.image.{Image, ImageView}
import scalafx.stage.FileChooser


/**
  * Created by Dragos on 6/14/2016.
  */
class AppController extends TAppController{

  override def continueExecute: PartialFunction[Any, Model] = {
    case (data: AppModel, Constants.addImage) => addImage(data)
    case (data: AppModel, Constants.clearCanvas) => clearCanvas(data)
  }

  override def addImage(data: AppModel): AppModel = {
    Util.runOnFxThread { new FileChooser().showOpenDialog(null) } match {
      case null => null
      case file: File =>
        val inputStream = new FileInputStream(file)

        try {

          val imageView = new ImageView()
          imageView.fitHeight = data.canvasHeight
          imageView.fitWidth = data.canvasWidth
          imageView.image = new Image(inputStream)

          AppModel(Constants.addImage, Util.runOnFxThread {imageView.snapshot(null,null)})
        } finally {
          inputStream.close()
        }
    }
  }

  override def clearCanvas(model: AppModel): AppModel = AppModel(Constants.clearCanvas, model.canvasWidth, model.canvasHeight)
}

File: src/main/scala-2.11/app/gui/controller/impl/AppController.java
package app.gui.controller.impl;

import app.gui.model.AppModelView;
import app.util.Command;
import javafx.beans.binding.Bindings;
import javafx.event.ActionEvent;
import javafx.fxml.Initializable;
import javafx.scene.canvas.Canvas;
import javafx.scene.control.Label;
import javafx.scene.control.Slider;
import javafx.scene.image.ImageView;
import javafx.scene.layout.GridPane;
import javafx.scene.layout.Pane;
import javafx.util.StringConverter;
import javafx.util.converter.NumberStringConverter;

import java.net.URL;
import java.util.ResourceBundle;

/**
 * Created by Dragos on 18.06.2016.
 */
public class AppController implements Initializable {
    public Slider redSlider;
    public Pane canvasPane;
    public Canvas canvas;
    public GridPane grid;
    public Label redSliderLabel;
    public ImageView inportedImage;
    public Slider greenSlider;
    public Label greenSliderLabel;
    public Slider blueSlider;
    public Label blueSliderLabel;
    public Slider opacitySlider;
    public Label opacitySliderLabel;
    public Slider contrastSlider;
    public Label contrastSliderLabel;
    public Slider hueSlider;
    public Label hueSliderLabel;
    public Slider brightnessSlider;
    public Label brightnessSliderLabel;
    public Slider saturationSlider;
    public Label saturationSliderLabel;

    private AppControllerScala appControllerScala;

    @Override
    public void initialize(URL location, ResourceBundle resources) {
        appControllerScala = new AppControllerScala(new AppModelView(
                Command.VIEW,
                canvas,
                canvasPane,
                grid,
                redSlider,
                redSliderLabel,
                greenSlider,
                greenSliderLabel,
                blueSlider,
                blueSliderLabel,
                opacitySlider,
                opacitySliderLabel,
                contrastSlider,
                contrastSliderLabel,
                hueSlider,
                hueSliderLabel,
                brightnessSlider,
                brightnessSliderLabel,
                saturationSlider,
                saturationSliderLabel,
                inportedImage
        ));

        StringConverter<Number> converter = new NumberStringConverter();
        Bindings.bindBidirectional(redSliderLabel.textProperty(), redSlider.valueProperty(), converter);
        Bindings.bindBidirectional(greenSliderLabel.textProperty(), greenSlider.valueProperty(), converter);
        Bindings.bindBidirectional(blueSliderLabel.textProperty(), blueSlider.valueProperty(), converter);
        Bindings.bindBidirectional(opacitySliderLabel.textProperty(), opacitySlider.valueProperty(), converter);

        Bindings.bindBidirectional(contrastSliderLabel.textProperty(), contrastSlider.valueProperty(), converter);
        Bindings.bindBidirectional(hueSliderLabel.textProperty(), hueSlider.valueProperty(), converter);
        Bindings.bindBidirectional(brightnessSliderLabel.textProperty(), brightnessSlider.valueProperty(), converter);
        Bindings.bindBidirectional(saturationSliderLabel.textProperty(), saturationSlider.valueProperty(), converter);

    }

    public void onImport_Click(ActionEvent actionEvent) {
        appControllerScala.onImport_Click(actionEvent);
    }

    public void onClear_Click(ActionEvent actionEvent) {
        appControllerScala.onClear_Click(actionEvent);
    }

    public void onReset_Click(ActionEvent actionEvent) {
        appControllerScala.onReset_Click(actionEvent);
    }
}
